### PR TITLE
Settings customizing tokenization

### LINKED ISSUE #2760: Internal representation of synonyms shouldn't be showed when displaying settings
## current behavior

When setting synonyms on the settings route:
```java
'推进之王': ['王维娜']
```
Synonyms are segmented and normalized before being stored in Meilisearch and the original value is dropped.
```java
{["推进", "之王"]: [["王维", "娜"]]}
```
> In this example, only the segmentation has an effect, but normalization could alter synonyms in higher proportions by replacing some characters. (`Dæmon` -> `daemon`)

When getting the settings (GET `/indexes/:uid/settings`) the displayed version of synonyms is the normalized one with 1 space between each segmented part.
```java
'推进 之王': ['王维 娜']
```

## drawback

This behavior leads to confusion because users don't see what they originally set.
related to: #2756
related to: https://github.com/meilisearch/product/discussions/503#discussioncomment-4086135
related to: #3883

## Possible enhancement

We should have both versions of the synonyms stored in different "databases"  to:
- remove the confusion coming from the inconsistency between what the user set and what is displayed
- allow Meilisearch to reformat internal synonyms enhancing the relevancy

### LINKED ISSUE #3610: Customize text separator
Related product team resources: [roadmap card](https://www.notion.so/meilisearch/ef7d86c80ef44c549e2361c70a367723?v=d11906bcef8b4b77815f84b801931c00&p=36d7f85c7eee465e950f92427e791738&pm=s)  (_internal only_)
Related spec: https://github.com/meilisearch/specifications/pull/254
Related product discussions:
- https://github.com/meilisearch/product/discussions/468
- https://github.com/meilisearch/product/discussions/160
- https://github.com/meilisearch/product/discussions/260
- https://github.com/meilisearch/product/discussions/381
- https://github.com/meilisearch/product/discussions/95

## Motivation

Adapts Meilisearch to specific term nomenclature and thus cover more use cases by adding customization to the engine.

Make Meilisearch more **customizable** and **adaptable** to different types of data set content and, therefore, different use cases by letting them configure the separators to be indexed.

More information present in the [product internal document](https://www.notion.so/meilisearch/ef7d86c80ef44c549e2361c70a367723?v=d11906bcef8b4b77815f84b801931c00&p=36d7f85c7eee465e950f92427e791738&pm=s).

## Usage

Refer to the final spec to know the detail of the usage.

## TODO

- [x] Implement changes in [Charabia](https://github.com/meilisearch/charabia/)
- [x] Release a Charabia version containing these changes
- [x] Bump this new Charabia version in Meilisearch and merge it into `main`
- [x] Add the new setting in Meilisearch to customize text separator (ensure that Meilisearch runs a reindexing)
- [x] Update the spec
- [x] ensure #3689 is fixed

## Impacted teams

@meilisearch/docs-team and @meilisearch/integration-team

### LINKED ISSUE #3917: Load a user-defined dictionary
Related product team resources: [roadmap card](https://www.notion.so/meilisearch/Send-my-own-word-dictionary-099bd4a69ef94614897728cdbc007110) (_internal only_)
Related product discussion:
- [Load a user-defined dictionary #663](https://github.com/meilisearch/product/discussions/663)
- https://github.com/meilisearch/meilisearch/issues/2879

Related spec: https://github.com/meilisearch/specifications/pull/255

## Motivation

Some Languages are not always space separated (Chinese, Japanese, Korean...) forcing Meilisearch to rely on common word dictionaries to segment a text into words, however, this strategy works only to segment common words and struggles when a text contains domain-specific words, technical words, or People Nouns.

Because it is impossible to have an exhaustive list of words of all the possible domains or a complete list of Nouns, Meilisearch should allow the user to add their own list of words specialized for their use case.

## Usage

TBC: Refer to the final spec to know the details and the final decisions about the usage.

## TODO

- [x] Release a prototype
- [ ] If prototype validated, merge changes into `main`
- [x] Update the spec

## Impacted teams

@meilisearch/docs-team @meilisearch/integration-team

### PR DESCRIPTION
**Pull request by ManyTheFish**
# Pull Request
This pull Request allows the User to customize Meilisearch Tokenization by providing specialized settings.

## Small documentation
All the new settings can be set and reset like the other index settings by calling the route `/indexes/:name/settings`

### `nonSeparatorTokens`
The Meilisearch word segmentation uses a default list of separators to segment words, however, for specific use cases some of the default separators shouldn't be considered separators, the `nonSeparatorTokens` setting allows to remove of some tokens from the default list of separators.

***Request payload `PUT`- `/indexes/articles/settings/non-separator-tokens`***
```json
["@", "#", "&"]
```

### `separatorTokens`
Some use cases need to define additional separators, some are related to a specific way of parsing technical documents some others are related to encodings in documents,  the `separatorTokens` setting allows adding some tokens to the list of separators.

***Request payload `PUT`- `/indexes/articles/settings/separator-tokens`***
```json
["&sect;", "&sep"]
```

### `dictionary`
The Meilisearch word segmentation relies on separators and language-based word-dictionaries to segment words, however, this segmentation is inaccurate on technical or use-case specific vocabulary (like `G/Box` to say `Gear Box`), or on proper nouns (like `J. R. R.` when parsing `J. R. R. Tolkien`), the `dictionary` setting allows defining a list of words that would be segmented as described in the list.

***Request payload `PUT`- `/indexes/articles/settings/dictionary`***
```json
["J. R. R.", "J.R.R."]
```

these last feature synergies well with the `stopWords` setting or the `synonyms` setting allowing to segment words and correctly retrieve the synonyms:
***Request payload `PATCH`- `/indexes/articles/settings`***
```json
{
    "dictionary": ["J. R. R.", "J.R.R."],
    "synonyms": {
            "J.R.R.": ["jrr", "J. R. R."],
            "J. R. R.": ["jrr", "J.R.R."],
            "jrr": ["J.R.R.", "J. R. R."],
    }
}
```

### Related specifications:
- https://github.com/meilisearch/specifications/pull/255
- https://github.com/meilisearch/specifications/pull/254

### Try it with Docker

```bash
$ docker pull getmeili/meilisearch:prototype-tokenizer-customization-3
```

## Related issue
Fixes #3610
Fixes #3917
Fixes https://github.com/meilisearch/product/discussions/468
Fixes https://github.com/meilisearch/product/discussions/160
Fixes https://github.com/meilisearch/product/discussions/260
Fixes https://github.com/meilisearch/product/discussions/381
Fixes https://github.com/meilisearch/product/discussions/131
Related to https://github.com/meilisearch/meilisearch/issues/2879

Fixes #2760

## What does this PR do?
- Add a setting `nonSeparatorTokens` allowing to remove a token from the default separator tokens
- Add a setting `separatorTokens` allowing to add a token in the separator tokens
- Add a setting `dictionary` allowing to override the segmentation on specific words
- add new error code `invalid_settings_non_separator_tokens` (invalid_request)
- add new error code `invalid_settings_separator_tokens` (invalid_request)
- add new error code `invalid_settings_dictionary` (invalid_request)

### CODE REVIEW

**Comment by tobiasnitsche** [comment]
Hi Gabi,

thanks for taking an eye and doing the effort, i really appreciate.

Two things came to mind:

- Would it be possible just to just use "PUT on /separatorTokens endpoint and to overwrite the whole setting? then i can just set "[]" to disable tokenization at all
- Is it possible to just use it on specific attributes? Similar to typoTolerance: https://www.meilisearch.com/docs/learn/configuration/typo_tolerance#disableonattributes . We have for example in our article index only one field "article-nr" which should have disabled the tokenization.

But thanks again for working on this!

**Comment by ManyTheFish** [comment]
Hello @tobiasnitsche,

First of all this feature is not meant to deactivate the tokenization in Meilisearch but to customize the behavior of the tokenizer. Deactivating the tokenization implies giving the user the full charge of it on the indexing side, the settings side and the search side, meaning that on each API we should accept an array of tokens instead of raw strings.


> - Would it be possible just to just use "PUT /separatorTokens" endpoint and to overwrite the whole tokenizer list to whats sent over? Then i could just set "[]" to disable tokenization at all.

No, you could eventually set the exhaustive list of separators in the `nonSeparatorTokens` list, but it wouldn't deactivate the tokenization, it would just avoid segmenting the words split by a separator contained in the list, and the other type of segmentation and the normalization wouldn't be deactivated.

> - Is it possible to just use it on specific attributes? Similar to typoTolerance: https://www.meilisearch.com/docs/learn/configuration/typo_tolerance#disableonattributes . We have for example in our article index only one attribute "serialnumber" which should have disabled the tokenization.

Not with this feature, the only thing you can do is using the `[disableonwords](https://www.meilisearch.com/docs/learn/configuration/typo_tolerance#disableonwords)` feature, but it's not constrained to a specific attribute.

Thank you for your report and your interest in the feature,
sorry if my answer is a bit deceptive,

see you!

**Comment by tobiasnitsche** [comment]
> Hello @tobiasnitsche,
> 
> First of all this feature is not meant to deactivate the tokenization in Meilisearch but to customize the behavior of the tokenizer. Deactivating the tokenization implies giving the user the full charge of it on the indexing side, the settings side and the search side, meaning that on each API we should accept an array of tokens instead of raw strings.
> 
> > * Would it be possible just to just use "PUT /separatorTokens" endpoint and to overwrite the whole tokenizer list to whats sent over? Then i could just set "[]" to disable tokenization at all.
> 
> No, you could eventually set the exhaustive list of separators in the `nonSeparatorTokens` list, but it wouldn't deactivate the tokenization, it would just avoid segmenting the words split by a separator contained in the list, and the other type of segmentation and the normalization wouldn't be deactivated.
> 
> > * Is it possible to just use it on specific attributes? Similar to typoTolerance: https://www.meilisearch.com/docs/learn/configuration/typo_tolerance#disableonattributes . We have for example in our article index only one attribute "serialnumber" which should have disabled the tokenization.
> 
> Not with this feature, the only thing you can do is using the `[disableonwords](https://www.meilisearch.com/docs/learn/configuration/typo_tolerance#disableonwords)` feature, but it's not constrained to a specific attribute.
> 
> Thank you for your report and your interest in the feature, sorry if my answer is a bit deceptive,
> 
> see you!

Thanks for your honest & detailed feedback on this!

File: milli/src/update/settings.rs [hunk 1]
```diff
@@ -452,9 +491,84 @@ impl<'a, 't, 'u, 'i> Settings<'a, 't, 'u, 'i> {
...
+                }
+            }
+            Setting::Reset => self.index.delete_non_separator_tokens(self.wtxn)?,
+            Setting::NotSet => false,
+        };
+
+        // the synonyms must be updated if non separator tokens have been updated.
+        if changes && self.synonyms == Setting::NotSet {
+            self.synonyms = Setting::Set(self.index.user_defined_synonyms(self.wtxn)?);
+        }
```
Comment (irevoire) [inline milli/src/update/settings.rs:514]: If there were no synonyms in the DB, shouldn’t we keep them as `NotSet`?

**Review by irevoire — CHANGES_REQUESTED** [review]
Nice, there are a lot of tests thanks!
I left a few questions, but overall, I think we'll be able to merge it in no time

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs [hunk 2]
```diff
@@ -52,6 +54,14 @@ pub fn extract_docid_word_positions<R: io::Read + io::Seek>(
     if let Some(stop_words) = stop_words {
         tokenizer_builder.stop_words(stop_words);
     }
+    if let Some(dictionary) = dictionary {
+        // let dictionary: Vec<_> = dictionary.iter().map(String::as_str).collect();
+        tokenizer_builder.words_dict(dictionary.as_slice());
+    }
+    if let Some(separators) = allowed_separators {
+        // let separators: Vec<_> = separators.iter().map(String::as_str).collect();
+        tokenizer_builder.separators(separators.as_slice());
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:63]: yes my bad

File: milli/src/update/settings.rs (hunk 1 above)
Comment (ManyTheFish) [reply milli/src/update/settings.rs:514]: The synonyms are changed only if there is a difference between the old version and the new one, when it's empty, nothing changes.

**Review by irevoire — APPROVED** [review]
Thanks!

bors merge

**Comment by tobiasnitsche** [comment]
> Hello @tobiasnitsche,
> 
> First of all this feature is not meant to deactivate the tokenization in Meilisearch...

Hi @ManyTheFish 

Actually this feature is not that useful for me, as the seperators can only be set on index base and not field base.

My index (in a nutshell)

articles:

article-number, description

A-12345 , "An article description"



To have nice searching experience, i need to set / disable the tokenizers only on the article number, the description should have the normal awesome meilisearch experience.

Is there any plan to add this feature on field base? Other settings like maxTypo has the option to set it on field base..

**Comment by curquiza** [comment]
Hello @tobiasnitsche 
thanks for your feedback, I let our product team know.

If you want to detail your usecase and directly talk with the product team, you can [open a discussion here](https://github.com/meilisearch/product/discussions) (or comment the already existing if there is one)

Thanks in advance for your feedback, it's really helpful

**Comment by tobiasnitsche** [comment]
Thank you @curquiza for encouraging comment. I love Meilisearch so far, keep on the good work!

**Comment by tobiasnitsche** [comment]
@curquiza @irevoire  

is there any update on this?

Actually for me , its a bit shameful, to tell the users, that they can not search for email addresses , or article numbers in the search...

Would love to have an update. This customization tokenization does not solve my problem, as i described it a few times...

**Comment by curquiza** [comment]
Hello @tobiasnitsche 

Sorry, but here it's a closed PR, we would rather avoid discussing too much in it to avoid losing track of information. PR are only for implementation and technical detail, not for product discussions 😊  Let's focus on the already existing support you interact in
- https://github.com/orgs/meilisearch/discussions/422
- https://github.com/meilisearch/meilisearch/discussions/3380

**Comment by tobiasnitsche** [comment]
Under stoo

> Hello @tobiasnitsche
> 
> Sorry, but here it's a closed PR, we would rather avoid discussing too much in it to avoid losing track of information. PR are only for implementation and technical detail, not for product discussions 😊 Let's focus on the already existing support you interact in
> 
>     * https://github.com/orgs/meilisearch/discussions/422
> 
>     * [Any idea to disable tokenization on specific field? #3380](https://github.com/meilisearch/meilisearch/discussions/3380)

Understood, lets move it there :-)

**Pull request merged**